File: twitch_integration.py
import requests
import os
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

class TwitchIntegration:
    def __init__(self, client_id, client_secret, oauth_token, broadcaster_id):
        self.client_id = client_id
        self.client_secret = client_secret
        self.oauth_token = oauth_token
        self.broadcaster_id = broadcaster_id
        self.base_url = "https://api.twitch.tv/helix"

    def _headers(self):
        return {
            "Client-ID": self.client_id,
            "Authorization": f"Bearer {self.oauth_token}",
        }

    def create_prediction(self, title, outcomes, duration):
        """
        Creates a Twitch prediction with the given title, outcomes, and duration.
        """
        url = f"{self.base_url}/predictions"
        data = {
            "broadcaster_id": self.broadcaster_id,
            "title": title,
            "outcomes": [{"title": outcome} for outcome in outcomes],
            "prediction_window": duration,
        }
        response = requests.post(url, json=data, headers=self._headers())
        if response.status_code == 201:
            print("Prediction created successfully.")
            return response.json()["data"][0]["id"]  # Return prediction ID
        else:
            print(f"Failed to create prediction: {response.status_code}, {response.text}")
            return None
            
    def get_active_predictions(self):
        """
        Fetches active predictions for the broadcaster.
        Returns a list of active predictions.
        """
        url = f"{self.base_url}/predictions?broadcaster_id={self.broadcaster_id}"
        response = requests.get(url, headers=self._headers())
        
        if response.status_code != 200:
            print(f"Failed to fetch predictions: {response.status_code}, {response.text}")
            return []

        try:
            predictions = response.json().get("data", [])
            active_predictions = [p for p in predictions if p["status"] == "ACTIVE" or p["status"] == "LOCKED"]

            if not active_predictions:
                print("No active predictions found.")
            return active_predictions
        except Exception as e:
            print(f"Error parsing prediction data: {e}")
            return []


    def end_prediction(self, winning_outcome_title=None):
        """
        Ends or cancels a prediction. If `winning_outcome_title` is None, cancels the prediction.
        """
        url = f"{self.base_url}/predictions"
        current_prediction = self.get_active_predictions()[0] if len(self.get_active_predictions()) == 1 else None
        yes_id = current_prediction["outcomes"][0]["id"]
        no_id = current_prediction["outcomes"][1]["id"]
        winning_outcome_id = yes_id if winning_outcome_title == 'Yes' else no_id
        
        data = {
            "broadcaster_id": self.broadcaster_id,
            "id": current_prediction['id'] if current_prediction else None,
            "status": "CANCELED" if winning_outcome_title is None else "RESOLVED",
            "winning_outcome_id": winning_outcome_id,
        }

        response = requests.patch(url, json=data, headers=self._headers())
        if response.status_code == 200:
            print("Prediction updated successfully.")
        else:
            print(f"Failed to update prediction: {response.status_code}, {response.text}")


    def run_ad(self, length):
        """
        Runs an ad on the Twitch channel for the given length in seconds.
        """
        url = f"{self.base_url}/channels/commercial"
        data = {
            "broadcaster_id": self.broadcaster_id,
            "length": length,
        }
        response = requests.post(url, json=data, headers=self._headers())
        if response.status_code == 200:
            print("Ad started successfully.")
        else:
            print(f"Failed to run ad: {response.status_code}, {response.text}")

# # Read credentials from environment variables
# CLIENT_ID = os.getenv("CLIENT_ID")
# CLIENT_SECRET = os.getenv("CLIENT_SECRET")
# OAUTH_TOKEN = os.getenv("OAUTH_TOKEN")
# BROADCASTER_ID = os.getenv("BROADCASTER_ID")

# twitch = TwitchIntegration(
#     client_id=CLIENT_ID,
#     client_secret=CLIENT_SECRET,
#     oauth_token=OAUTH_TOKEN,
#     broadcaster_id=BROADCASTER_ID
# )

# twitch.end_prediction('Yes')

File: TwitchIntegration.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using System.Linq;

namespace TwitchPredictionTest
{
    public class TwitchIntegration
    {
        private readonly string _clientId;
        private readonly string _clientSecret;
        private readonly string _oauthToken;
        private readonly string _broadcasterId;
        private readonly string _baseUrl = "https://api.twitch.tv/helix";

        private JsonSerializerOptions _jsonOptions = new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true
        };

        public TwitchIntegration(string clientId, string clientSecret, string oauthToken, string broadcasterId)
        {
            _clientId = clientId;
            _clientSecret = clientSecret;
            _oauthToken = oauthToken;
            _broadcasterId = broadcasterId;
        }

        private HttpClient CreateHttpClient()
        {
            var client = new HttpClient();
            client.DefaultRequestHeaders.Add("Client-ID", _clientId);
            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _oauthToken);
            return client;
        }

        public async Task<string> CreatePredictionAsync(string title, List<string> outcomes, int duration)
        {
            using (var client = CreateHttpClient())
            {
                var url = $"{_baseUrl}/predictions";
                var data = new
                {
                    broadcaster_id = _broadcasterId,
                    title = title,
                    outcomes = outcomes.ConvertAll(outcome => new { title = outcome }),
                    prediction_window = duration
                };

                var json = JsonSerializer.Serialize(data, _jsonOptions);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var response = await client.PostAsync(url, content);
                var responseContent = await response.Content.ReadAsStringAsync();
                Console.WriteLine($"Create prediction response: {responseContent}"); // Debug line

                if (response.IsSuccessStatusCode)
                {
                    var predictionResponse = JsonSerializer.Deserialize<PredictionResponse>(responseContent, _jsonOptions);
                    if (predictionResponse?.Data != null && predictionResponse.Data.Count > 0)
                    {
                        Console.WriteLine("Prediction created successfully.");
                        return predictionResponse.Data[0].Id;
                    }
                }
                
                Console.WriteLine($"Failed to create prediction: {response.StatusCode}, {responseContent}");
                return null;
            }
        }

        public async Task<List<Prediction>> GetActivePredictionsAsync()
        {
            using (var client = CreateHttpClient())
            {
                var url = $"{_baseUrl}/predictions?broadcaster_id={_broadcasterId}";
                var response = await client.GetAsync(url);

                if (!response.IsSuccessStatusCode)
                {
                    Console.WriteLine($"Failed to fetch predictions: {response.StatusCode}, {await response.Content.ReadAsStringAsync()}");
                    return new List<Prediction>();
                }

                try
                {
                    var responseJson = await response.Content.ReadAsStringAsync();
                    Console.WriteLine($"Raw prediction response: {responseJson}"); // Debug line
                    var predictionResponse = JsonSerializer.Deserialize<PredictionResponse>(responseJson, _jsonOptions);
                    
                    if (predictionResponse?.Data == null)
                    {
                        Console.WriteLine("No predictions data returned");
                        return new List<Prediction>();
                    }

                    var activePredictions = predictionResponse.Data
                        .Where(p => p.Status == "ACTIVE" || p.Status == "LOCKED")
                        .ToList();

                    if (activePredictions.Count == 0)
                    {
                        Console.WriteLine("No active predictions found.");
                    }
                    else
                    {
                        Console.WriteLine($"Found {activePredictions.Count} active predictions");
                    }

                    return activePredictions;
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error parsing prediction data: {ex.Message}");
                    return new List<Prediction>();
                }
            }
        }

        public async Task EndPredictionAsync(string predictionId, string winningOutcomeId = null)
        {
            using (var client = CreateHttpClient())
            {
                var url = $"{_baseUrl}/predictions";
                var data = new
                {
                    broadcaster_id = _broadcasterId,
                    id = predictionId,
                    status = winningOutcomeId == null ? "CANCELED" : "RESOLVED",
                    winning_outcome_id = winningOutcomeId
                };

                var json = JsonSerializer.Serialize(data);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var response = await client.PatchAsync(url, content);
                if (response.IsSuccessStatusCode)
                {
                    Console.WriteLine("Prediction updated successfully.");
                }
                else
                {
                    Console.WriteLine($"Failed to update prediction: {response.StatusCode}, {await response.Content.ReadAsStringAsync()}");
                }
            }
        }

        public async Task RunAdAsync(int length)
        {
            using (var client = CreateHttpClient())
            {
                var url = $"{_baseUrl}/channels/commercial";
                var data = new
                {
                    broadcaster_id = _broadcasterId,
                    length = length
                };

                var json = JsonSerializer.Serialize(data);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var response = await client.PostAsync(url, content);
                if (response.IsSuccessStatusCode)
                {
                    Console.WriteLine("Ad started successfully.");
                }
                else
                {
                    Console.WriteLine($"Failed to run ad: {response.StatusCode}, {await response.Content.ReadAsStringAsync()}");
                }
            }
        }

        public class PredictionResponse
        {
            public List<Prediction> Data { get; set; }
        }

        public class Prediction
        {
            public string Id { get; set; }
            public string BroadcasterId { get; set; }
            public string BroadcasterName { get; set; }
            public string BroadcasterLogin { get; set; }
            public string Title { get; set; }
            public string WinningOutcomeId { get; set; }
            public string Status { get; set; }
            public int Duration { get; set; }
            public DateTime CreatedAt { get; set; }
            public List<Outcome> Outcomes { get; set; }
        }

        public class Outcome
        {
            public string Id { get; set; }
            public string Title { get; set; }
            public int Users { get; set; }
            public int ChannelPoints { get; set; }
            public string Color { get; set; }
        }
    }
}